src/libraries/Native/Unix/System.Native/pal_networking.c

```diff
@@ -432,6 +422,137 @@ int32_t SystemNative_GetHostEntryForName(const uint8_t* address, HostEntry* entr
     return ret;
 }
 
+#if HAVE_GETADDRINFO_A
+struct GetAddrInfoAsyncState
+{
+    struct gaicb gai_request;
+    struct gaicb* gai_requests;
+    struct sigevent sigevent;
+
+    const uint8_t* address;
+    HostEntry* entry;
+    GetHostEntryForNameCallback callback;
+};
+
+static void GetHostEntryForNameAsyncComplete(sigval_t context)
+{
+    struct GetAddrInfoAsyncState* state = (struct GetAddrInfoAsyncState*)context.sival_ptr;
+
+    atomic_thread_fence(memory_order_acquire);
```

Comment:
Why is the fence necessary?

Reply:
This callback comes in on another thread. It ensures visibility to this thread of everything in `state`.

`getaddrinfo_a` probably does something like this already, so it might be safe to remove, but it isn't documented.